godot-core/src/builtin/array.rs

```diff
@@ -126,13 +111,120 @@ impl Array {
+        self.as_inner().sort();
+    }
+
+    /// Shuffles the array such that the items will have a random order. This method uses the
+    /// global random number generator common to methods such as `randi`. Call `randomize` to
+    /// ensure that a new seed will be used each time if you want non-reproducible shuffling.
+    pub fn shuffle(&mut self) {
+        self.as_inner().shuffle();
+    }
+
+    /// Asserts that the given index refers to an existing element.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn check_bounds(&self, index: usize) {
+        let len = self.len();
+        assert!(
+            index < len,
+            "Array index {index} is out of bounds: length is {len}",
+        );
+    }
+
+    /// Returns a pointer to the element at the given index.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn ptr(&self, index: usize) -> *const Variant {
+        self.check_bounds(index);
+        // SAFETY: We just checked that the index is not out of bounds.
+        let ptr = unsafe { self.ptr_unchecked(index) };
+        assert!(!ptr.is_null());
+        ptr
+    }
+
+    /// Returns a mutable pointer to the element at the given index.
+    ///
+    /// # Panics
+    ///
+    /// If `index` is out of bounds.
+    fn ptr_mut(&self, index: usize) -> *mut Variant {
+        self.check_bounds(index);
+        // SAFETY: We just checked that the index is not out of bounds.
+        let ptr = unsafe { self.ptr_mut_unchecked(index) };
+        assert!(!ptr.is_null());
+        ptr
+    }
+
+    /// Returns a pointer to the element at the given index.
+    ///
+    /// # Safety
+    ///
+    /// Calling this with an out-of-bounds index is undefined behavior.
+    unsafe fn ptr_unchecked(&self, index: usize) -> *const Variant {
+        let item_ptr: sys::GDExtensionVariantPtr =
+            (interface_fn!(array_operator_index_const))(self.sys(), to_i64(index));
+        item_ptr as *const Variant
+    }
+
+    /// Returns a mutable pointer to the element at the given index.
+    ///
+    /// # Safety
+    ///
+    /// Calling this with an out-of-bounds index is undefined behavior.
+    unsafe fn ptr_mut_unchecked(&self, index: usize) -> *mut Variant {
+        let item_ptr: sys::GDExtensionVariantPtr =
+            (interface_fn!(array_operator_index))(self.sys(), to_i64(index));
+        item_ptr as *mut Variant
+    }
+
+    #[doc(hidden)]
+    pub fn as_inner(&self) -> inner::InnerArray {
+        inner::InnerArray::from_sys_ptr(self.sys())
+    }
+
+    /// Changes the generic type on this array, without changing its contents. Needed for API
+    /// functions that return a variant array even though we know its type, and for API functions
+    /// that take a variant array even though we want to pass a typed one.
+    ///
+    /// This is marked `unsafe` since it breaks the invariant that a `TypedArray<T>` always holds a
+    /// Godot array whose runtime type is `T`.
```

Comment:
Would need a `# Safety` section.

Also, "since it breaks" -> "since it can be used to break".
Hopefully it doesn't _always_ break 😉

Reply:
Added.